### DISCUSSION TITLE
Parallel calculation with 4 process in GNSS processing

### DISCUSSION THREAD

**Discussion by FFFFFG-IGG** [groops-devs/groops#116] [Q&A] [answered]
Execuse me, developers.
  I've gotten your kind advice in #107 to improve efficiency by using MPI. However, when I set processCountPerIteration=1 in my loopProgram in 01preprocessing.xml and 03Network.xml and run them with"mpiexec -n 4 groopsMPI"(under ubuntu), I got different results:
1. In the case of 01Preprocessing.xml(a loopProgram contains 7-days iteration), parallel log shows that 3 processes are used and(process 3 day simultineously), and the efficiency is much more better than no-parallel.
2. In the case of 03Network.xml(a loopProgram also contains 7-days iteration),parallel log shows that only one process is used(process one day simultineously), the efficiency doesn't improve significanly.
![b853e0f256ec06f848a49ac5fadbdca](https://user-images.githubusercontent.com/59271400/195051102-e66e01e0-b6c6-4d0f-96aa-3fca93561283.jpg)
Note that my computer's CPU has 8 cores with 32 running-cache, I hope to get some help from you developers to solve what happened in 03Network.xml, and the  01Preprocessing.xml,03Network.xml are in the attachment.
[xml.zip](https://github.com/groops-devs/groops/files/9753555/xml.zip)
Best regards,
Yongzhao Fan.

**Comment by tmayerguerr** [comment, marked as answer]
Dear Yongzhao Fan,

From the documentation of [LoopProgram](https://groops-devs.github.io/groops/html/LoopPrograms.html)

> If this program is executed on multpile processing nodes, the iterations can be computed in parallel, see parallelization. The first process serves as load balancer and the other processes are assigned to iterations according to processCountPerIteration. For example, running a loop containing three iterations on 13 processes with processCountPerIteration=4, runs the three iterations in parallel, with each iteration being assigned four processes.

With your settings the outer loop is parallelized and the loop steps are computed in parallel. Therefore the programs inside the loop are executed only by processCountPerIteration=1 process each. This is also true for the inner [LoopProgram](https://groops-devs.github.io/groops/html/LoopPrograms.html), so no additional parallelization is possible.

Set processCountPerIteration=0 for the outer loop or maybe you can merge the two [LoopProgram](https://groops-devs.github.io/groops/html/LoopPrograms.html) by using [Loop](https://groops-devs.github.io/groops/html/loopType.html)->loop

regards
Torsten
